linkerd/config/src/main/scala/io/buoyant/linkerd/config/Parser.scala

```diff
@@ -0,0 +1,50 @@
+package io.buoyant.linkerd.config
+
+import com.fasterxml.jackson.core.JsonFactory
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer
+import com.fasterxml.jackson.databind.module.SimpleModule
+import com.fasterxml.jackson.databind.{DeserializationFeature, DeserializationContext, ObjectMapper}
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper
+import io.buoyant.linkerd.config.types._
+import scala.reflect.{ClassTag, classTag}
+
+abstract class ConfigDeserializer[T: ClassTag] extends StdDeserializer[T](Parser.jClass[T]) {
+  def register(module: SimpleModule): SimpleModule = module.addDeserializer(Parser.jClass[T], this)
+  protected def catchMappingException(ctxt: DeserializationContext)(t: => T): T =
+    try t catch { case arg: IllegalArgumentException => throw ctxt.mappingException(arg.getMessage) }
+}
+
+object Parser {
+  def jClass[T: ClassTag]: Class[T] = classTag[T].runtimeClass.asInstanceOf[Class[T]]
```

Comment:
it might be good to add a comment indicating why we need to do this typecasting (i.e. erasure)